FFmpeg: Bump to 3.0

fritsch opened the pull request:
This bumps ffmpeg to version 3.0 and tries to fix all warnings.

Commit:
FFmpeg: windows bump

fritsch committed:
FFMPEG: Bump depends version to 3.0 branch

fritsch committed:
FFmpeg: Replace libdcadec with internal hd aware dca decoder

fritsch committed:
Demuxers / Codec: Request dcadec instead of libdcadec

fritsch committed:
FFmpeg 3.0 Transition: Use av_packet_unref instead of av_free_packet

fritsch committed:
FFmpeg 3.0 Transition: Use AVPicture to AVFrame transition

fritsch commented:
@wsnipex Could you please check if I have cut out the libdcadec stuff correctly?

@FernetMenta and @popcornmix please have a look especially at the AVPicture to AVFrame transition.

@afedchin Thanks for the Windows commit - anything missing for you?

jenkins build this please

afedchin commented:
if we drop external dcadec please pick this up https://github.com/afedchin/xbmc/commit/0c8b045dd5444d9f186b944d6ef88274d4607dd2

fritsch commented:
@afedchin already in: https://github.com/xbmc/xbmc/pull/9491/commits/e7d53f07edd0db78d3c8c3975d3391f609d20975 right?

afedchin commented:
oh. misread. that's right

fritsch committed:
FFmpeg 3.0 Transition: DVDOverlayCodecFFmpeg - use linesize and data directly

wm4 commented on xbmc/guilib/FFmpegImage.cpp:
```diff
@@ -375,17 +375,17 @@ bool CFFmpegImage::Decode(unsigned char * const pixels, unsigned int width, unsi
 bool CFFmpegImage::DecodeFrame(AVFrame* frame, unsigned int width, unsigned int height, unsigned int pitch, unsigned char * const pixels)
 {
-  AVPicture* pictureRGB = static_cast<AVPicture*>(av_mallocz(sizeof(AVPicture)));
+  AVFrame* pictureRGB = static_cast<AVFrame*>(av_mallocz(sizeof(AVFrame)));
```
This is slightly broken. sizeof(AVFrame) is not part of the ABI. Use av_frame_alloc() (if I remember the function name right).

fritsch replied:
Will do - thx much!

fritsch commented:
@ace20022 Please also have a look at our picture class. Thanks much.

FernetMenta commented on xbmc/video/FFmpegVideoDecoder.cpp:
```diff
@@ -47,23 +47,23 @@ FFmpegVideoDecoder::~FFmpegVideoDecoder()
-  if ( m_pFrame )
-	av_free( m_pFrame );
+  if (m_pFrame)
+    av_free(m_pFrame);
```
you can trash the entire class. it is not used anymore

fritsch committed:
FFmpegVideoDecoder: Trash entire class

fritsch committed:
FFmpeg 3.0 Transition: duration is int64_t now too - workaround can be dropped

wsnipex commented on tools/depends/target/ffmpeg/autobuild.sh:
```diff
@@ -163,7 +163,6 @@ CFLAGS="$CFLAGS" CXXFLAGS="$CXXFLAGS" LDFLAGS="$LDFLAGS" \
 	--disable-mips32r2 \
 	--disable-mipsdspr1 \
 	--disable-mipsdspr2 \
```
according to travis, these options seem to have been removed
Unknown option "--disable-mipsdspr1".

fritsch replied:
Thx, fixed.

fritsch committed:
FFmpeg 3.0 Transition: Alter changed mips compile params

fritsch committed:
FFmpeg 3.0 Transition: Use 16 byte alignment

fritsch commented:
@stefansaraev can you retry with the latest 16 Byte alignment?

fritsch commented:
jenkins build this please

fritsch commented:
@Memphiz 

```
Undefined symbols for architecture armv7:
  "_VTDecompressionSessionCreate", referenced from:
      _av_videotoolbox_default_init2 in libavcodec.a(videotoolbox.o)
  "_VTDecompressionSessionDecodeFrame", referenced from:
      _videotoolbox_common_end_frame in libavcodec.a(videotoolbox.o)
  "_VTDecompressionSessionInvalidate", referenced from:
      _av_videotoolbox_default_free in libavcodec.a(videotoolbox.o)
  "_VTDecompressionSessionWaitForAsynchronousFrames", referenced from:
      _videotoolbox_common_end_frame in libavcodec.a(videotoolbox.o)
ld: symbol(s) not found for architecture armv7
```

we need some explicit --enable-videotoolbox for IOS?

fritsch committed:
IOS: explicitely disable videotoolbox (for now)

fritsch commented:
jenkins build this please for ios

wsnipex commented:
this is a linking issue. Probably some vtd libs that ffmpeg uses not linked against kodi.

fritsch commented:
@wsnipex yes - see my appended commit. VTB is now default in ffmpeg, but we use something home baked.

MrMC commented:
@fritsch Much worse :)

tvOS:
platform_min_version="$target_platform-version-min=9.0"

iOS:
platform_min_version="$target_platform-version-min=5.1"

VTB is prohibited under tvOS and was private under iOS until iOS8. 

DVDVideoCodecVideoToolBox.cpp looks for and dyloads it.

So must be disabled in ffmpeg for now.

fritsch commented:
@MrMc we will use it as default in the future from ffmpeg, but for now I disabled it.

fritsch committed:
FFmpeg 3.0 Transition: warn when external supplied buffers are not properly aligned

Background: when ffmpeg is build with asm intrinsics memory adresses need proper
alignment

fritsch committed:
Texture: Use 16 byte alignment

fritsch commented:
I added another commit for Texture to use our _aligned_malloc method to properly align. When using ffmpeg's ASM methods this will crash and burn when we get external Textures supplied and our buffers are not 16 aligned.

Let's see if it builds for all of them.

jenkins build this please

ace20022 commented:
This one https://github.com/xbmc/xbmc/pull/9491/commits/27cdbd53bb25f4c799b586055390d4360cb459b4 is completely unrelated

fritsch commented:
@ace20022 that's right somehow I worked on the wrong branch :-)

fritsch commented:
jenkins build this please

fritsch commented:
jenkins is happy. Windows yellow is something else.

chovy commented:
This breaks any addon (ie: xbmc-gamepass or nba-leagupass official addons) that play streams from neulion.

fritsch commented:
File a bug at trac.kodi.tv, please. Until now you are the very first and only that reports those issues, though without _any_ debuglogs at all.

fritsch commented on xbmc/cores/VideoPlayer/DVDCodecs/Audio/DVDAudioCodecFFmpeg.cpp:
```diff
@@ -65,7 +65,7 @@ bool CDVDAudioCodecFFmpeg::Open(CDVDStreamInfo &hints, CDVDCodecOptions &options
       allowdtshddecode = atoi(it->m_value.c_str());
 
   if (hints.codec == AV_CODEC_ID_DTS && allowdtshddecode)
-    pCodec = avcodec_find_decoder_by_name("libdcadec");
```
Afaik on pi1 the dvdcodecoptions were set at that time. Let's remove it and see.

FernetMenta replied:
the point there is no option anymore to allow dtdhd or are you aware of any?  explicitly selecting dcadec for DTS is pointless because it is the only and default codec for dts.

fritsch replied:
Wasnt it in pi1 specific settings that @popcornmix carries? Let's wait for his comment.

FernetMenta replied:
what do you intent do do with a setting, that can't change anything? again, ffmpeg provides no means for selecting different codecs for dts. hence the setting in this part of the code is obsolete.

fritsch replied:
``` for(std::vector<CDVDCodecOption>::iterator it = options.m_keys.begin(); it != options.m_keys.end(); ++it)
     if (it->m_name == "allowdtshddecode")
allowdtshddecode = atoi(it->m_value.c_str());
 ```
